content/codewars-python-test-framework-v2.md

```diff
@@ -0,0 +1,298 @@
+
+<br>
+<br>
+
+You may find here the detailed information about the [test framework](https://github.com/codewars/python-test-framework/blob/master/codewars_test/test_framework.py) designed for CodeWars. The present version is extending the first one, so old katas are still compatible, but the use of the v2 is required for newly created katas or translations.
+
+_Note: the test module is aliased with `test`, so you can call every tool provided with it using either `Test.method` or `test.method`._
+
+<br>
+<br>
+
+---
+
+# Overview
+
+The test framework allowes to write named groups of tests holding other named groups of tests, containing themselves assertions.
+
+* Assertions (using the methods detailed below) can actually be used at any level in the groups, or even outise of them (but you're advised to avoid that).
+* Note that the code executed inside on group isn't stopped at the first failed assertion (unlees you use `allow_raise`. See below), so the full test suite is executed each time.
+
+The basic setup for the tests follow this example:
+
+```python
+@Test.describe('Fixed Tests')
+def example_tests():
+
+    @Test.it('Example Test Case')
+    def example_test_case():
+        test.assert_equals(add(1, 1), 2, 'Optional Message on Failure')
+    
+    @Test.it("More tests")
+    def more():
+        for a,b,exp in [(-2,30,28), (42,0,42)]:
+            Test.assert_equals(add(a,b), exp)
+            
+    @Test.it("Pair tests")
+    def more():
+        for v in range(10):
+            Test.assert_equals(add(v,v), 2*v)
+```
+
+Here is the result visible by the user in the output panel:
+
+`TODO: add the image: "docs/content/PythonTestFrameWorkExample.png"  => how?`
+
+
+
+<br>
+<br>
+
+
+
+
+## Tests grouping (blocks)
+
+The groups are created using the following decorators:
+
+```python
+@Test.describe(test_name)
+def _():
+
+    @Test.it(test_case_name)
+    def _():
+        ...
+         
+    @Test.describe(subgroup_name)
+    def _():
+      ...
+
+        @Test.it(test_case_subgroup_name)
+        def _():
+            ...
+```
+
+Those decorators automatically run the decorated function to launch the tests. Both print elapsed time of the block at the end, and `describe`s can be nested.
+
+
+* `describe` blocks:
+
+Those blocks always stay depolyed whatever the results of the tests are, menaing that if you put assertions directly in them, they will be visible in the output, as well as any other `it` block(s) it could contain. They are the basic block inwhich you can insert `it` blocks or other nested `describe` blocks if you need to create subdivisions in your test suite.
+
+* `it` blocks:
+
+On the contrary of the `describe` blocks, the `it` blocks are deployed in the output panel only when they contain failed assertions. Oterwise they are shrinked and only the name of the block and the number of assertions are visible.
+
+Note that you _cannot nest_ the `it` blocks.
+
+
+
+<br>
+<br>
+
+
+
+---
+
+
+# Tests interruption
```

Comment:
Only use `h1` (`# heading`) once for the title of the document.

Reply:
yep, I made the correction.

Reply:
? I meant once at the top `# Python test framework for Codewars (v2)`. There are still this one and `# Available assertion methods`. Those should be `##` and any children increased by one level.

Reply:
mmh? I thought I already corrected that one too. I'll check